data-prepper-api/src/main/java/org/opensearch/dataprepper/model/source/SourceCoordinator.java

```diff
@@ -0,0 +1,56 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.dataprepper.model.source;
+
+import java.time.Duration;
+import java.util.List;
+import java.util.Optional;
+
+/**
+ * Interface to be passed to {@link org.opensearch.dataprepper.model.source.Source} to enable distribution and coordination of work
+ * between source plugins, and to track progress and save state for these partitions of work
+ * @since 2.2
+ */
+public interface SourceCoordinator {
```

Comment:
This can be generic which would make development easier for sources.

This snippet shows the change that would make it generic.
```
public interface SourceCoordinator<T> {
  Optional<SourcePartition<T>> getNextPartition();
  void saveStateForPartition(final String partitionKey, final T partitionState);
}
```